### ISSUE TITLE
How to filter bad result?

### ISSUE THREAD

**Issue by Javacr** [facebookresearch/InterWild#17]
Hello, Moon! I don't know how to filter out the bad results, because as long as I give InterWild the hand bbox, it always returns the key points. I tried using the confidence provided by hand detector , but it wasn't reliable(as shown in the following figs

). I wanted to use the confidence of joint_img, but InterWild didn't provide it. So, can you give me some advice?
![image](https://github.com/facebookresearch/InterWild/assets/44012990/e2af7d48-d49f-4c7b-b6bb-7a9b1b5d3113)

![image](https://github.com/facebookresearch/InterWild/assets/44012990/5e48898a-529c-428f-88d3-174f75a30772)

![image](https://github.com/facebookresearch/InterWild/assets/44012990/8cefc282-bc80-4893-b0b7-5169cb287a81)

**Comment by mks0601** [comment]
Hi, could you use this one?
https://github.com/facebookresearch/InterWild/blob/70d3eca941270f103cefec8907b9b62590321ed6/demo/demo.py#L112

**Comment by Javacr** [comment]
> Hi, could you use this one?
> 
> https://github.com/facebookresearch/InterWild/blob/70d3eca941270f103cefec8907b9b62590321ed6/demo/demo.py#L112

I've actually tried hand_bbox_conf, but it's not reliable. In fig 1, the palm is blocked, but InterWild gives high confidences; In fig 3, palm is clear, but gives low confidence. I decide to add a parallel network to obtain 2D hand joints and corresponding confidence, so that judge the quality of 3D results.

**Comment by mks0601** [comment]
Actually, you can do the same things as here (https://github.com/facebookresearch/InterWild/blob/70d3eca941270f103cefec8907b9b62590321ed6/common/nets/module.py#L225) to here (https://github.com/facebookresearch/InterWild/blob/70d3eca941270f103cefec8907b9b62590321ed6/common/nets/module.py#L31) to get joint confidence. Could you try?

**Comment by Javacr** [comment]
> Actually, you can do the same things as here (
> 
> https://github.com/facebookresearch/InterWild/blob/70d3eca941270f103cefec8907b9b62590321ed6/common/nets/module.py#L225
> 
> ) to here (
> https://github.com/facebookresearch/InterWild/blob/70d3eca941270f103cefec8907b9b62590321ed6/common/nets/module.py#L31
> 
> ) to get joint confidence. Could you try?

Thank you for your suggestion, I have tried it today.
 **module.py**
```
class PositionNet(nn.Module):
    # 2.5D joint
    def __init__(self):
        super(PositionNet, self).__init__()
        self.joint_num = mano.sh_joint_num
        self.depth_dim = cfg.output_hand_hm_shape[0]
        self.conv = make_conv_layers([2048, self.joint_num*self.depth_dim], kernel=1, stride=1, padding=0, bnrelu_final=False)

    def get_conf(self, hand_hm, rhand_joint, lhand_joint):
        # rhand_joint: Size([21, 3])
        batch_size, joint_num, depth, height, width = hand_hm.shape
        hand_hm = hand_hm.view(batch_size, joint_num, depth*height*width)
        hand_hm = F.softmax(hand_hm, 2)
        hand_hm = hand_hm.view(batch_size, joint_num, depth, height, width)
        # hand_hm: Size([2, 21, 8, 8, 8])
        rjoint_conf = sample_joint_features_3d(hand_hm[0,::], rhand_joint)
        ljoint_conf = sample_joint_features_3d(hand_hm[1,::], lhand_joint)
        return rjoint_conf, ljoint_conf
    
    def forward(self, hand_feat):
        hand_hm = self.conv(hand_feat)
        _, _, height, width = hand_hm.shape
        hand_hm = hand_hm.view(-1,self.joint_num,self.depth_dim,height,width)
        # import pdb; pdb.set_trace()
        # hand_coord: Size([2, 21, 3])
        hand_coord = soft_argmax_3d(hand_hm)
        rhand_joint, lhand_joint = hand_coord[0,:,:], hand_coord[1,:,:]
        rjoint_conf, ljoint_conf = self.get_conf(hand_hm, rhand_joint, lhand_joint)
        return hand_coord, rjoint_conf, ljoint_conf
```
 **transforms.py**
```
def sample_joint_features_3d(heatmap3d, joint_xyz):
    # heatmap3d: Size([21, 8, 8, 8]) -> Size([21, 1, 8, 8, 8])
    # joint_xyz: Size([21, 3])-> Size(21, 1, 3)
    # grid: Size([21, 1, 1, 1, 3])
    heatmap3d = heatmap3d[:,None,:,:,:]
    joint_xyz = joint_xyz[:,None,:]
    depth, height, width = heatmap3d.shape[2:]
    # normalize to [-1, 1]
    x = joint_xyz[:,:,0] / (width-1) * 2 - 1
    y = joint_xyz[:,:,1] / (height-1) * 2 - 1
    z = joint_xyz[:,:,2] / (depth-1) * 2 - 1

    grid = torch.stack((x,y,z),2)[:,:,None,None,:]
    # import pdb; pdb.set_trace()
    # input math:`(N, C, D_\text{in}, H_\text{in}, W_\text{in})` (5-D case)
    # grid math:`(N, D_\text{out}, H_\text{out}, W_\text{out}, 3)` (5-D case)
    confidence_joint = F.grid_sample(heatmap3d, grid, align_corners=True)[:,:,:,0] # batch_size, channel_dim, joint_num
    return confidence_joint
```
I give some results(all joint confidences are show on image), two hands is clear in left image, but blurry in right image. 
case 1: The mean confidences are higher than 0.08 in left image and lower than 0.08 in righ image, this is reasonable.
![image](https://github.com/facebookresearch/InterWild/assets/44012990/5c6489c3-50d3-4a03-a193-a706234617cb)

case 2: The mean confidence is higher than 0.08 in left image, this is reasonable, howeve, the mean confidence of right hand in right image is 0.1, it is too high.
![image](https://github.com/facebookresearch/InterWild/assets/44012990/ed20fc92-7a03-488d-85b0-b373f15d6ba3)

case 3: like case2
![image](https://github.com/facebookresearch/InterWild/assets/44012990/4f6489e5-f598-48bd-8ad2-ff0ae5bb47e6)

**Comment by Javacr** [comment]
I remove softmax when I obtain confidence, this means I use ture value of  hand_hm to represent confidence. When value is smaller than 0, 0 is used instead. The new results seem reasonable.
![image](https://github.com/facebookresearch/InterWild/assets/44012990/0f1f3f71-bd9a-4a53-8faa-84c4a6c76e8a)
![image](https://github.com/facebookresearch/InterWild/assets/44012990/a33f6156-373d-4132-848e-cc79374ce013)
![image](https://github.com/facebookresearch/InterWild/assets/44012990/06b3c59f-a4b5-42dc-ab60-67e68329bd07)

**Comment by yanqi1811** [comment]
Hi, Javacr! Is your hands detetion from interwild model or you have retrained a hands detection model? I have tested interwild model and I found the detection results are not good when whole body in image. Thanks!

**Comment by Javacr** [comment]
> Hi, Javacr! Is your hands detetion from interwild model or you have retrained a hands detection model? I have tested interwild model and I found the detection results are not good when whole body in image. Thanks!

Could you paste some results? The hand detector is from interwild, it preform well in most case. Its shortcoming is that it will predict two hands if only one hand in image, so I use wrist joint from VIT pose to filter out the wrong bbox.

**Comment by yanqi1811** [comment]
> > Hi, Javacr! Is your hands detetion from interwild model or you have retrained a hands detection model? I have tested interwild model and I found the detection results are not good when whole body in image. Thanks!
> 
> Could you paste some results? The hand detector is from interwild, it preform well in most case. Its shortcoming is that it will predict two hands if only one hand in image, so I use wrist joint from VIT pose to filter out the wrong bbox.

Like these. These iamges resolution are 1920*1080, I think this resolution is very large, maybe I should run a body detection firstly. 
 
![20231102174143](https://github.com/facebookresearch/InterWild/assets/31723725/47409a90-b482-4105-bf84-cad52d5607cb)
![20231102174724](https://github.com/facebookresearch/InterWild/assets/31723725/dbd33956-4436-4c80-ad60-3f779059409e)

**Comment by Javacr** [comment]
> > > Hi, Javacr! Is your hands detetion from interwild model or you have retrained a hands detection model? I have tested interwild model and I found the detection results are not good when whole body in image. Thanks!
> > 
> > 
> > Could you paste some results? The hand detector is from interwild, it preform well in most case. Its shortcoming is that it will predict two hands if only one hand in image, so I use wrist joint from VIT pose to filter out the wrong bbox.
> 
> Like these. These iamges resolution are 1920*1080, I think this resolution is very large, maybe I should run a body detection firstly. When I crop the body image, it works well. Thank you!
> 
> ![20231102174143](https://user-images.githubusercontent.com/31723725/279936546-47409a90-b482-4105-bf84-cad52d5607cb.jpg) ![20231102174724](https://user-images.githubusercontent.com/31723725/279936554-dbd33956-4436-4c80-ad60-3f779059409e.jpg) ![20231102180356](https://user-images.githubusercontent.com/31723725/279940634-15ed4363-d584-4088-8d64-3d72866a2156.jpg)

As moon said, "Put input images at images. The image should be a cropped image, which contain a single human. For example, using a human detector. We have a hand detection network, so no worry about the hand postiions!", human detector is important.

**Comment by yanqi1811** [comment]
> > > > Hi, Javacr! Is your hands detetion from interwild model or you have retrained a hands detection model? I have tested interwild model and I found the detection results are not good when whole body in image. Thanks!
> > > 
> > > 
> > > Could you paste some results? The hand detector is from interwild, it preform well in most case. Its shortcoming is that it will predict two hands if only one hand in image, so I use wrist joint from VIT pose to filter out the wrong bbox.
> > 
> > 
> > Like these. These iamges resolution are 1920*1080, I think this resolution is very large, maybe I should run a body detection firstly. When I crop the body image, it works well. Thank you!
> > ![20231102174143](https://user-images.githubusercontent.com/31723725/279936546-47409a90-b482-4105-bf84-cad52d5607cb.jpg) ![20231102174724](https://user-images.githubusercontent.com/31723725/279936554-dbd33956-4436-4c80-ad60-3f779059409e.jpg) ![20231102180356](https://user-images.githubusercontent.com/31723725/279940634-15ed4363-d584-4088-8d64-3d72866a2156.jpg)
> 
> As moon said, "Put input images at images. The image should be a cropped image, which contain a single human. For example, using a human detector. We have a hand detection network, so no worry about the hand postiions!", human detector is important.

Thank you!
